File: old/617-merge-bintree/js/617.js
#!/usr/bin/node
const TreeNode = (val) => ({
    val: val,
    left: null,
    right: null,
});

let mergeTrees = function(t1, t2) {
    if (t1 == null) {
        return t2;
    }
    if (t2 == null) {
        return t1;
    }
    t1.val += t2.val;
    t1.left = mergeTrees(t1.left, t2.left);
    t1.right = mergeTrees(t1.right, t2.right);
    return t1;
};

let et1 = TreeNode(1);
et1.left = TreeNode(3);
et1.left.left = TreeNode(5);
et1.right = TreeNode(2);

let et2 = TreeNode(2);
et2.left = TreeNode(1);
et2.left.right = TreeNode(4);
et2.right = TreeNode(3);
et2.right.right = TreeNode(7);

let actualResult = mergeTrees(et1, et2);
console.log(actualResult);

File: old/617-merge-bintree/elixir/617.exs
#! /usr/bin/env elixir
defmodule BinTree do
  defstruct value: nil, left: nil, right: nil

  def merge(nil, nil), do: nil
  def merge(%BinTree{} = t1, nil), do: t1
  def merge(nil, %BinTree{} = t2), do: t2

  def merge(%BinTree{} = t1, %BinTree{} = t2) do
    %BinTree{
      value: t1.value + t2.value,
      left: merge(t1.left, t2.left),
      right: merge(t1.right, t2.right)
    }
  end
end

defmodule BinTreeTest do
  defp bt(value, left, right), do: %BinTree{value: value, left: left, right: right}
  defp leaf(value), do: %BinTree{value: value}

  defp t1, do: bt(1, bt(3, leaf(5), nil), leaf(2))
  defp t2, do: bt(2, bt(1, nil, leaf(4)), bt(3, nil, leaf(7)))

  defp t_expected, do: bt(3, bt(4, leaf(5), leaf(4)), bt(5, nil, leaf(7)))

  def go do
    (BinTree.merge(t1(), t2()) == t_expected())
    |> IO.inspect()
  end
end

BinTreeTest.go()
